File: tests/algorithms/x/PHP/dynamic_programming/rod_cutting.php
<?php
error_reporting(E_ALL & ~E_DEPRECATED);
ini_set('memory_limit', '-1');
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
function _panic($msg) {
    fwrite(STDERR, strval($msg));
    exit(1);
}
function enforce_args($n, $prices) {
  if ($n < 0) {
  _panic('n must be non-negative');
}
  if ($n > count($prices)) {
  _panic('price list is shorter than n');
}
}
function bottom_up_cut_rod($n, $prices) {
  enforce_args($n, $prices);
  $max_rev = null;
  $i = 0;
  while ($i <= $n) {
  if ($i == 0) {
  $max_rev = _append($max_rev, 0);
} else {
  $max_rev = _append($max_rev, -2147483648);
}
  $i = $i + 1;
};
  $length = 1;
  while ($length <= $n) {
  $best = $max_rev[$length];
  $j = 1;
  while ($j <= $length) {
  $candidate = $prices[$j - 1] + $max_rev[$length - $j];
  if ($candidate > $best) {
  $best = $candidate;
}
  $j = $j + 1;
};
  $max_rev[$length] = $best;
  $length = $length + 1;
};
  return $max_rev[$n];
}
$prices = [1, 5, 8, 9, 10, 17, 17, 20, 24, 30];
echo rtrim(json_encode(bottom_up_cut_rod(4, $prices), 1344)), PHP_EOL;
echo rtrim(json_encode(bottom_up_cut_rod(10, $prices), 1344)), PHP_EOL;

File: tests/algorithms/transpiler/FS/dynamic_programming/rod_cutting.fs
// Generated 2025-08-13 07:12 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let _arrset (arr:'a array) (i:int) (v:'a) : 'a array =
    let mutable a = arr
    if i >= a.Length then
        let na = Array.zeroCreate<'a> (i + 1)
        Array.blit a 0 na 0 a.Length
        a <- na
    a.[i] <- v
    a
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let rec enforce_args (n: int) (prices: int array) =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    let mutable n = n
    let mutable prices = prices
    try
        if n < 0 then
            ignore (failwith ("n must be non-negative"))
        if n > (Seq.length (prices)) then
            ignore (failwith ("price list is shorter than n"))
        __ret
    with
        | Return -> __ret
and bottom_up_cut_rod (n: int) (prices: int array) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable n = n
    let mutable prices = prices
    try
        enforce_args (n) (prices)
        let mutable max_rev: int array = Array.empty<int>
        let mutable i: int = 0
        while i <= n do
            if i = 0 then
                max_rev <- Array.append max_rev [|0|]
            else
                max_rev <- Array.append max_rev [|int (-2147483648L)|]
            i <- i + 1
        let mutable length: int = 1
        while length <= n do
            let mutable best: int = _idx max_rev (int length)
            let mutable j: int = 1
            while j <= length do
                let candidate: int = (_idx prices (int (j - 1))) + (_idx max_rev (int (length - j)))
                if candidate > best then
                    best <- candidate
                j <- j + 1
            max_rev.[length] <- best
            length <- length + 1
        __ret <- _idx max_rev (int n)
        raise Return
        __ret
    with
        | Return -> __ret
let prices: int array = unbox<int array> [|1; 5; 8; 9; 10; 17; 17; 20; 24; 30|]
ignore (printfn "%d" (bottom_up_cut_rod (4) (prices)))
ignore (printfn "%d" (bottom_up_cut_rod (10) (prices)))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)
